include/nbl/video/utilities/IGPUObjectFromAssetConverter.h

```diff
@@ -855,76 +864,117 @@ auto IGPUObjectFromAssetConverter::create(const asset::ICPUImage** const _begin,
+        barrier.subresourceRange.baseArrayLayer = 0u;
+        barrier.subresourceRange.layerCount = cpuimg->getCreationParameters().arrayLayers;
+
+        asset::SImageBlit blitRegion = {};
+        blitRegion.srcSubresource.aspectMask = barrier.subresourceRange.aspectMask;
+        blitRegion.srcSubresource.baseArrayLayer = barrier.subresourceRange.baseArrayLayer;
+        blitRegion.srcSubresource.layerCount = barrier.subresourceRange.layerCount;
+        blitRegion.srcOffsets[0] = { 0, 0, 0 };
+
+        blitRegion.dstSubresource.aspectMask = barrier.subresourceRange.aspectMask;
+        blitRegion.dstSubresource.baseArrayLayer = barrier.subresourceRange.baseArrayLayer;
+        blitRegion.dstSubresource.layerCount = barrier.subresourceRange.layerCount;
+        blitRegion.dstOffsets[0] = { 0, 0, 0 };
+
+        // Compute mips
+        auto mipsize = cpuimg->getMipSize(cpuimg->getCreationParameters().mipLevels);
+        uint32_t mipWidth = mipsize.x;
+        uint32_t mipHeight = mipsize.y;
+        uint32_t mipDepth = mipsize.z;
+        for (uint32_t i = cpuimg->getCreationParameters().mipLevels; i < gpuimg->getCreationParameters().mipLevels; ++i)
         {
-            tmpViewParams.flags = static_cast<IGPUImageView::E_CREATE_FLAGS>(0u);
-            tmpViewParams.image = core::smart_refctd_ptr<IGPUImage>(img);
-            switch (img->getCreationParameters().type)
-            {
-            case asset::IImage::ET_1D:
-                tmpViewParams.viewType = IGPUImageView::ET_1D_ARRAY;
-                break;
-            case asset::IImage::ET_2D:
-                if (img->getCreationParameters().flags & asset::IImage::ECF_CUBE_COMPATIBLE_BIT)
-                    tmpViewParams.viewType = IGPUImageView::ET_CUBE_MAP_ARRAY;
-                else
-                    tmpViewParams.viewType = IGPUImageView::ET_2D_ARRAY;
-                break;
-            case asset::IImage::ET_3D:
-                tmpViewParams.viewType = IGPUImageView::ET_3D;
-                break;
-            default:
-                assert(false);
-                break;
-            }
-            tmpViewParams.format = img->getCreationParameters().format;
-            //tmpViewParams.subresourceRange.aspectMask
-            tmpViewParams.subresourceRange.baseMipLevel = lowestPresentMip - 1u;
-            tmpViewParams.subresourceRange.layerCount = img->getCreationParameters().arrayLayers;
-            auto tmpView = _params.device->createGPUImageView(std::move(tmpViewParams));
-
-            // deprecated OpenGL path (do with compute shader in the future)
-            cmdbuf_compute->regenerateMipmaps(tmpView.get());
+            const uint32_t srcLoD = i - 1u;
+
+            barrier.barrier.srcAccessMask = asset::EAF_TRANSFER_WRITE_BIT;
+            barrier.barrier.dstAccessMask = asset::EAF_TRANSFER_READ_BIT;
+            barrier.oldLayout = asset::EIL_TRANSFER_DST_OPTIMAL;
+            barrier.newLayout = asset::EIL_TRANSFER_SRC_OPTIMAL;
```

Comment:
a blit requires a transfer SRC layout!?

Really!? @achalpandeyy @Erfan-Ahmadi ?

Reply:
1.
> VUID-vkCmdBlitImage-srcImageLayout-01398
srcImageLayout must be VK_IMAGE_LAYOUT_SHARED_PRESENT_KHR, VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL or VK_IMAGE_LAYOUT_GENERAL
> dstImage must have been created with VK_IMAGE_USAGE_TRANSFER_DST_BIT usage flag
2. 
> VK_PIPELINE_STAGE_TRANSFER_BIT specifies the following commands:
All copy commands, including vkCmdCopyQueryPoolResults
vkCmdBlitImage2KHR and vkCmdBlitImage
vkCmdResolveImage2KHR and vkCmdResolveImage
All clear commands, with the exception of vkCmdClearAttachments
https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineStageFlagBits.html

Reply:
lets transfer the entire image to `VK_IMAGE_LAYOUT_GENERAL` once at the beginning after the region upload (instead of SHADER_READ_ONLY), less PITA and less code. Then SHADER_READ_ONLY at the end of mipmapping.

Furthermore it will already be as its needed to be for the compute shader mipmapping.